superbench/benchmarks/micro_benchmarks/cublas_function/cublas_benchmark.h

```diff
@@ -195,95 +219,240 @@ class CublasFunction {
-    // input argument
-    CUDA_SAFE_CALL(cudaMallocHost((void **)&Parameter_1_0_host, sizeof(cuComplex) * n * k * this->batch_count_));
-    CUDA_SAFE_CALL(cudaMalloc((void **)Parameter_1_0, sizeof(cuComplex) * n * k * this->batch_count_));
-
-    // fill input values
-    fill_data_cucomplex(Parameter_0_0_host, Parameter_1_0_host);
+template <typename T> T *CublasFunction::transpose(const T *matrix, int m, int n, int batch_count) {
+    T *transpose_matrix = (T *)malloc((unsigned long)m * (unsigned long)n * sizeof(T) * (unsigned long)batch_count);
+    for (int b = 0; b < batch_count; b++) {
+        for (int i = 0; i < m * n; i++) {
+            int c = i / m;
+            int r = i % m;
+            int tran_i = r * n + c;
+            transpose_matrix[tran_i + b * m * n] = matrix[i + b * m * n];
+        }
+    }
+    return transpose_matrix;
+}
+/**
+ * @brief Matrix multiply calculation on CPU side
+ */
+template <typename T1, typename T2>
+void CublasFunction::matrix_calculation_on_cpu_with_data(const T1 *Parameter_0_0_host, const T1 *Parameter_1_0_host,
+                                                         const T1 *Result_3_0, T2 **Result_cpu, T2 alpha, T2 beta) {
+    int m = this->m_, n = this->n_, k = this->k_, batch_count = this->batch_count_;
+    // Copy result from device to host
+    T1 *Result_3_0_host;
+    CUDA_SAFE_CALL(cudaMallocHost((void **)&Result_3_0_host,
+                                  sizeof(T1) * (unsigned long)m * (unsigned long)n * (unsigned long)batch_count));
+    CUDA_SAFE_CALL(cudaMemcpy(Result_3_0_host, Result_3_0,
+                              sizeof(T1) * (unsigned long)m * (unsigned long)n * (unsigned long)batch_count,
+                              cudaMemcpyDeviceToHost));
+    // Transpose the input matrix
+    T1 *Parameter_0_0_host_op, *Parameter_1_0_host_op;
+    Parameter_0_0_host_op = (T1 *)malloc((unsigned long)m * (unsigned long)k * sizeof(T1) * (unsigned long)batch_count);
+    Parameter_1_0_host_op = (T1 *)malloc((unsigned long)n * (unsigned long)k * sizeof(T1) * (unsigned long)batch_count);
+    memcpy(Parameter_0_0_host_op, Parameter_0_0_host,
+           (unsigned long)m * (unsigned long)k * sizeof(T1) * (unsigned long)batch_count);
+    memcpy(Parameter_1_0_host_op, Parameter_1_0_host,
+           (unsigned long)n * (unsigned long)k * sizeof(T1) * (unsigned long)batch_count);
+    if (this->transa_) {
+        Parameter_0_0_host_op = transpose(Parameter_0_0_host, k, m, batch_count);
+    }
+    if (this->transb_) {
+        Parameter_1_0_host_op = transpose(Parameter_1_0_host, n, k, batch_count);
+    }
+    // C + i*strideC = alpha*op(A+i*strideA)*op(B+i*strideB)+beta(C+i*strideC), for i in [0, batchcount -1 ]
+    // reference in https://docs.nvidia.com/cuda/cublas/index.html#cublas-level-3-function-reference
+    *Result_cpu = (T2 *)malloc((unsigned long)m * (unsigned long)n * sizeof(T2) * (unsigned long)batch_count);
```

Comment:
Will `Result_cpu` will be malloc more than once before free it?

Reply:
no, the malloc in this function and free in check_result function are paired